pilot/cmd/istioctl/controlplane/cmd.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2017 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controlplane
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+func Command(istioNamespaceFlag *string) *cobra.Command {
+	var features *[]string
+	var out string
+
+	m := defaultModel()
+	cmd := &cobra.Command{
+		Use:     "control-plane",
+		Short:   "Generates the configuration for Istio's control plane.",
+		Long:    `istioctl control-plane produces deployment files to run the minimum Istio control for the set of features requested by the --feature flag. If no features are provided, we create deployments for the default control plane: Pilot, Mixer, CA, and Ingress Proxies, with mTLS enabled.`,
+		Example: `istioctl control-plane --features routing,policy,initializer -o helm`,
+		RunE: func(c *cobra.Command, args []string) error {
+			m.setFeatures(*features)
+			m.Namespace = *istioNamespaceFlag
+			switch strings.ToLower(out) {
+			case "helm":
+				_, err := fmt.Fprint(os.Stdout, fromModel(m))
+				return err
+			case "yaml":
+				return fmt.Errorf("not implemented")
+			default:
+				return fmt.Errorf("unsupported output %q", out)
+			}
+			return nil
+		},
+	}
+
+	features = cmd.PersistentFlags().StringArrayP("features", "f", []string{},
+		`List of Istio features to enable. Accepts any combination of "mtls", "telemetry", "routing", "ingress", "policy", "initializer".`)
+	cmd.PersistentFlags().StringVar(&m.Hub, "hub", "gcr.io/istio-testing", "The container registry to pull Istio images from")
+	cmd.PersistentFlags().StringVar(&m.MixerTag, "mixer-tag", "latest", "The tag to use to pull the `mixer` container")
```

Comment:
let's not have more than 1 tag ? (or at least have 1 flag setting the 1 tag to all that live in the same source tree)

Reply:
We use different tags for each of them today:

https://github.com/istio/istio/blob/c679bceb1308917e1b4e87e4dc3d6a77ce2e1a6c/install/kubernetes/helm/istio/values.yaml#L21-L26

Reply:
Also a super common use-case is that I want to test a custom image I just pushed. It'd be nice to just `istioctl control-plane --pilot-tag zbutcher` to update the configs to use my new version. I want to add hub-per-component for the same reason, so I can use normal everything but refer to my own test hub.

Reply:
yes single component dev is one use case, but "check everything I may have changed" (ie build all from source and deploy that) is too, I guess it can be a convenience on top

We probably have to figure out a way to bake into the binary or read from some place the hub/tags to use so they match what was built (and have it work for releases with release tags and hubs too)

Reply:
We have to be careful not to complicate the end-user by adding support for these nice to have dev options.

Reply:
That's fair - in general I don't expect those options will be used frequently, though I think we'll want them if we use this tool to generate the YAML files we ship Istio with.

How about I mark all of the hub/tag flags as hidden so they don't show up in the help text, but they'll still exist for our own use cases?

Reply:
Can we group them in the help description, and specify clearly they are optional and should be used to override images to custom values?

Reply:
AFAIK cobra doesn't provide an easy way to group together a set of flags (short of using subcommands) and always prints the flags in alphabetical order. I marked them as hidden so they won't show up in help text but are still usable if desired.